base/applications/rapps/geninst.cpp

```diff
@@ -0,0 +1,813 @@
+        if (err)
+            ErrorBox(err);
+    }
+}
+
+static BOOL
+AddEntry(WCHAR Type, LPCWSTR Value)
+{
+    InstallInfo &Info = *static_cast<InstallInfo *>(g_pInfo);
+    CStringW name;
+    name.Format(L"%c%u", Type, ++Info.EntryCount);
+    UINT err = Info.Entries.SetStringValue(name, Value);
+    if (err)
+        ErrorBox(err);
+    return !err;
+}
+
+static HRESULT
+GetCustomIconPath(InstallInfo &Info, CStringW &Path)
+{
+    if (*GetGenerateString(DB_GENINST_ICON, Path))
+    {
+        Path = BuildPath(Info.InstallDir, Path);
+        int idx = PathParseIconLocation(Path.GetBuffer());
+        Path.ReleaseBuffer();
+        HICON hIco = NULL;
+        if (ExtractIconExW(Path, idx, &hIco, NULL, 1))
+            DestroyIcon(hIco);
+        if (idx)
+            Path.AppendFormat(L",%d", idx);
+        return hIco ? S_OK : S_FALSE;
+    }
+    return HRESULT_FROM_WIN32(ERROR_NO_MORE_ITEMS);
+}
+
+static BOOL
+GetLocalizedSMFolderName(LPCWSTR WinVal, LPCWSTR RosInf, LPCWSTR RosVal, CStringW &Output)
+{
+    CRegKey key;
+    if (key.Open(HKEY_LOCAL_MACHINE, L"Software\\Microsoft\\Windows\\CurrentVersion", KEY_READ) == ERROR_SUCCESS &&
+        GetRegString(key, WinVal, Output) && !Output.IsEmpty())
+    {
+        return TRUE;
+    }
+    WCHAR windir[MAX_PATH];
+    GetWindowsDirectoryW(windir, _countof(windir));
+    CStringW path = BuildPath(BuildPath(windir, L"inf"), RosInf), section;
+    DWORD lang = 0, lctype = LOCALE_ILANGUAGE | LOCALE_RETURN_NUMBER;
+    if (GetLocaleInfoW(GetUserDefaultLCID(), lctype, (LPWSTR) &lang, sizeof(lang) / sizeof(WCHAR)))
+    {
+        section.Format(L"Strings.%.4x", lang);
+        if (ReadIniValue(path, section, RosVal, Output) > 0)
+            return TRUE;
+        section.Format(L"Strings.%.2x", PRIMARYLANGID(lang));
+        if (ReadIniValue(path, section, RosVal, Output) > 0)
+            return TRUE;
+    }
+    return ReadIniValue(path, L"Strings", RosVal, Output) > 0;
+}
+
+static BOOL
+CreateShortcut(const CStringW &Target)
+{
+    InstallInfo &Info = *static_cast<InstallInfo *>(g_pInfo);
+    UINT csidl = Info.PerUser ? CSIDL_PROGRAMS : CSIDL_COMMON_PROGRAMS;
+    CStringW rel = Info.ShortcutFile, path, dir, tmp;
+
+    if (FAILED(GetSpecialPath(csidl, path, Info.GetGuiOwner())))
+        return TRUE; // Pretend everything is OK
+
+    int cat;
+    if (Info.Parser.GetInt(DB_CATEGORY, cat) && cat == ENUM_CAT_GAMES)
+    {
+        if (GetLocalizedSMFolderName(L"SM_GamesName", L"shortcuts.inf", L"Games", tmp))
+        {
+            path = BuildPath(path, tmp);
+        }
+    }
+
+    rel.Replace('/', '\\');
+    path = BuildPath(path, rel.GetString());
+    UINT SHPPFW = SHPPFW_DIRCREATE | SHPPFW_IGNOREFILENAME;
+    HRESULT hr = SHPathPrepareForWriteW(Info.GetGuiOwner(), NULL, path, SHPPFW);
+    if ((Info.Error = ErrorFromHResult(hr)) != 0)
+    {
+        ErrorBox(Info.Error);
+        return FALSE;
+    }
+
+    IShellLinkW *pSL;
```

Comment:
Please use `CComPtr` for this.

Reply:
It's less than 20 lines before it's released again but OK.